api/resolvers/sub.js

```diff
@@ -315,6 +315,56 @@ export default {
       notifyTerritoryTransfer({ models, sub, to: user })
 
       return updatedSub
+    },
+    unarchiveTerritory: async (parent, { name, hash, hmac, ...data }, { me, models, lnd }) => {
+      if (!me) {
+        throw new GraphQLError('you must be logged in', { extensions: { code: 'UNAUTHENTICATED' } })
+      }
+
+      const oldSub = await models.sub.findUnique({ where: { name } })
+      if (!oldSub) {
+        throw new GraphQLError('sub not found', { extensions: { code: 'BAD_INPUT' } })
+      }
+      if (oldSub.status !== 'STOPPED') {
+        throw new GraphQLError('sub is not archived', { extensions: { code: 'BAD_INPUT' } })
+      }
+      if (oldSub.billingType === 'ONCE') {
+        // sanity check. this should never happen but leaving this comment here
+        // to stop error propagation just in case and document that this should never happen.
+        // #defensivecode
+        throw new GraphQLError('sub should not be archived', { extensions: { code: 'BAD_INPUT' } })
+      }
```

Comment:
I'd like us to always reuse the exact same validation logic on the server as the client. We used to do them separately but they tended to get out of sync which would lead to loopholes in validation.

Reply:
Mhh, syncing client and server validation here is a bit tricky because I am using the same schema for two different mutations and switch between the used mutation based on the validation result in the client :thinking: 

Probably a sign that I shouldn't be doing this but we don't have a pattern for validation warnings yet.

But yes, this is missing the validation of the other fields like in `upsertSub` so `ssValidate(territorySchema, ...)` should be used here in some way.

Reply:
I am using `onChange` now. Much better. To not prevent submission, I added logic as was done for edits to validation. See eb8b7e08. There is some double negation going on since we want to check for the non-existence and then exclude some territories from that check, but I think it's understandable enough.

I have also added server validation via `ssValidate(territorySchema, ...)` in 723c7969. I have kept the other checks since they still make sense as they are not caught during validation for frontend / control flow reasons:

1. Sub must exist to unarchive
2. Sub must be archived to unarchive
3. Sub should have been billed once